plugins/ssr/ssrfwd.ml

```diff
@@ -318,3 +318,130 @@ let sufftac ist ((((clr, pats),binders),simpl), ((_, c), hint)) =
+      | Term.ProdType _ -> aux t
+      | Term.AtomicType(hd, args) when
+          Ssrcommon.is_ind_ref sigma hd (Coqlib.lib_ref "core.eq.type") &&
+          Array.length args = 3 && is_app_evar sigma args.(2) ->
+            Tactics.New.refine ~typecheck:true (fun sigma ->
+              let sigma, under =
+                Ssrcommon.mkSsrConst "Under" env sigma in
+              let sigma, under_from_eq =
+                Ssrcommon.mkSsrConst "Under_from_eq" env sigma in
+              let ty = EConstr.mkApp (under,args) in
+              let sigma, t = Evarutil.new_evar env sigma ty in
+              sigma, EConstr.mkApp(under_from_eq,Array.append args [|t|]))
+      | _ -> Proofview.tclUNIT ()
+  in
+    aux c
+
+end
+
+let rec pretty_rename evar_map term = function
+  | [] -> term
+  | varname :: varnames ->
+     try begin match EConstr.destLambda evar_map term with (_, types, body) ->
+           let res = pretty_rename evar_map body varnames in
+           let name = Names.Name.mk_name varname in
+           (* Note: we don't explicitly check name \notin free_vars(res) here *)
+           EConstr.mkLambda (name, types, res)
+         end
+     with
+     | DestKO ->
+        ppdebug(lazy Pp.(str"under: cannot pretty-rename all bound variables with destLambda"));
+        term
+
+let overtac gl = ssr_n_tac "over" ~-1 gl
+
+let check_numgoals ?(minus = 0) nh =
+  Proofview.numgoals >>= fun ng ->
+  if nh <> ng then
+    let errmsg =
+      str"Incorrect number of hints" ++ spc() ++
+        str"(expected "++int (ng - minus)++str(String.plural ng " tactic") ++
+        str", was given "++ int (nh - minus)++str")."
+    in
+    CErrors.user_err errmsg
+  else
+    Proofview.tclUNIT ()
+
+let undertac ist varnames ((dir,mult),_ as rule) hint =
+  if mult <> Ssrequality.nomult then
+    Ssrcommon.errorstrm Pp.(str"Multiplicity not supported");
+
+  let map_redex env evar_map ~before:_ ~after:t =
+    ppdebug(lazy Pp.(str("under vars: " ^
+      (List.fold_right (fun e r -> Names.Id.to_string e ^ " " ^ r) varnames ""))));
+    ppdebug(lazy Pp.(str"under: mapping:" ++ pr_econstr_env env evar_map t));
+
+    let evar_map, ty = Typing.type_of env evar_map t in
+    let new_t = (* pretty-rename the bound variables *)
+      try begin match EConstr.destApp evar_map t with (f, ar) ->
+            let lam = Array.last ar in
+            let new_lam = pretty_rename evar_map lam varnames in
+            let new_ar, len1 = Array.copy ar, pred (Array.length ar) in
+            new_ar.(len1) <- new_lam;
+            EConstr.mkApp (f, new_ar)
+          end with
+      | DestKO ->
+        ppdebug(lazy Pp.(str"under: cannot pretty-rename bound variables with destApp"));
+        t
+    in
+    ppdebug(lazy Pp.(str"under: to:" ++ pr_econstr_env env evar_map new_t));
+    evar_map, new_t
+  in
+  let undertacs =
+    if hint = nohint then
+      Proofview.tclUNIT ()
+    else
+      let simpl = Ssrequality.simpltac (Simpl ~-1) in
```

Comment:
There must be a watic to perform `nf_betaiota`, it is done almost systematically after every `refine` but I can't find it now.

Reply:
@gares Taking some inspiration from
https://github.com/coq/coq/blob/d31056ec924ef6e09b28bc3822b427b67a8a300b/plugins/ltac/rewrite.ml#L1567
I've just addressed this in a commit, but only with `nf_beta` which appears to suffice…